pkg/email/email_test.go

```diff
@@ -0,0 +1,216 @@
+		{name: "missing_recipient", smtpHost: "test@test.com", sender: "alice@test.com", receiver: []string{}, expected: email.ErrNoRecipientConfigured},
+		{name: "missing_smtp_host_and_sender", smtpHost: "", sender: "", receiver: []string{"bob@test.com"}, expected: email.ErrNoSMTPHostConfigured},
+		{name: "missing_smtp_host_and_recipient", smtpHost: "", sender: "alice@test.com", receiver: []string{}, expected: email.ErrNoSMTPHostConfigured},
+		{name: "missing_all", smtpHost: "", sender: "", receiver: []string{}, expected: email.ErrNoSMTPHostConfigured},
+		{name: "missing_sender_and_recipient", smtpHost: "test@test.com", sender: "", receiver: []string{}, expected: email.ErrNoSenderConfigured},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			emailer.Params.SMTPHost, emailer.Params.Sender = tt.smtpHost, tt.sender
+			err = emailer.SendEmail(tt.receiver, "bar", "baz", nil)
+			if !errors.Is(err, tt.expected) {
+				t.Errorf("expected err: %s got err: %s", tt.expected, err)
+			}
+		})
+	}
+}
+
+func TestSendEmail_Headers(t *testing.T) {
+	p := email.Params{
+		SMTPHost: "abc@test.com",
+		Sender:   "test@test.com",
+	}
+	d := Dialer{}
+	emailer := email.Emailer{
+		Params: p,
+		Dialer: &d,
+	}
+	const subject = "testSubject"
+	const receiver = "receiver@test.com"
+	err := emailer.SendEmail([]string{receiver}, subject, "", nil)
+	testutil.Must(t, err)
+	if len(d.Message) != 1 {
+		t.Fatalf("message length %d, expected 1", len(d.Message))
+	}
+	headers := map[string]string{
+		"From":    p.Sender,
+		"To":      receiver,
+		"Subject": subject,
+	}
+	for k, v := range headers {
+		header := d.Message[0].GetHeader(k)[0]
+		if header != v {
+			t.Errorf("got header: %s, expected header: %s", header, v)
+		}
+	}
+}
+
+func TestSendEmailWithLimit(t *testing.T) {
+	p := email.Params{SMTPHost: "test.com", Sender: "alice@testing.com"}
+	emailer, err := email.NewEmailer(p)
+	testutil.Must(t, err)
+	emailer.Dialer = &Dialer{}
+	tests := []struct {
+		name       string
+		limitEvery time.Duration
+		burst      int
+		expect     error
+	}{
+		{name: "normal", limitEvery: rate.InfDuration, burst: math.MaxInt, expect: nil},
+		{name: "zero_burst", limitEvery: rate.InfDuration, burst: 0, expect: email.ErrRateLimitExceeded},
+		{name: "zero_time", limitEvery: 0, burst: math.MaxInt, expect: nil},
+		{name: "zero_burst_and_time", limitEvery: 0, burst: 0, expect: email.ErrRateLimitExceeded},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			emailer.Limiter = rate.NewLimiter(rate.Limit(tt.limitEvery), tt.burst)
+			err := emailer.SendEmailWithLimit([]string{"receiver@test.com"}, "foo", "", nil)
+			if !errors.Is(err, tt.expect) {
+				t.Errorf("got err: %s, expected err %s", err, tt.expect)
+			}
+		})
+	}
+}
+
+func TestSendResetPasswordAndInviteUserEmail_SendWithLimit(t *testing.T) {
```

Comment:
You are not checking SendWithLimit in this content

Reply:
If I understood you correctly, the issue here is the function name? (Since the underlying code calls the SendEmailWithLimit as written in the comment)
Changed the function name